contracts/multi-token/src/lib.rs

```diff
@@ -132,14 +171,178 @@ impl SimpleMTKCore for SimpleMTK {
     /// * sender MUST have sufficient amount of token.
     ///
     /// Arguments:
-    /// * `id`: Token ID.
-    /// * `amount`: Token's amount to be burnt.
-    fn burn(&mut self, id: TokenId, amount: u128) {
-        MTKCore::burn(self, vec![id], vec![amount]);
-        let sup = self.supply(id);
-        let mut _balance = self
-            .supply
-            .insert(self.token_id, sup.saturating_sub(amount));
+    /// * `ids`: The vector of token IDs
+    /// * `amounts`: The vector of token amounts to be burnt.
+
+    fn burn(&mut self, ids: Vec<TokenId>, amounts: Vec<u128>) -> Result<MtkEvent, MtkError> {
+        if ids.len() != amounts.len() {
+            return Err(MtkError::LengthMismatch);
+        }
+
+        ids.iter()
+            .zip(amounts.clone())
+            .try_for_each(|(id, amount)| {
+                if self.tokens.token_metadata.contains_key(id) && amount > 1 {
+                    return Err(MtkError::AmountGreaterThanOneForNft);
+                }
+                self.check_opportunity_burn(&msg::source(), id, amount)
+            })?;
+
+        ids.iter()
+            .enumerate()
+            .for_each(|(i, id)| self.burn_impl(id, amounts[i]));
+
+        for (id, amount) in ids.iter().zip(amounts.iter()) {
+            let quantity = self.supply.get_mut(id).unwrap();
+            *quantity = quantity.saturating_sub(*amount);
+        }
+
+        Ok(MtkEvent::Transfer {
+            from: msg::source(),
```

Comment:
`msg::source()` is called several times (also in `burn_impl`)
it's necessary to save `msg::source()` value into variable